numpy/core/function_base.py

```diff
@@ -159,14 +180,20 @@ def linspace(start, stop, num=50, endpoint=True, retstep=False, dtype=None,
         # Multiply with delta to allow possible override of output class.
         y = y * delta
 
-    y += start
+    if is_datelike:
+        y = y + start
+    else:
+        y += start
 
     if endpoint and num > 1:
         y[-1] = stop
 
     if axis != 0:
         y = _nx.moveaxis(y, 0, axis)
 
+    if is_datelike:
+        y = y.astype(dtype)
```

Comment:
Is this line still necessary? I wonder if it makes more sense to move the `astype` up to cast `start` already (or at least put it into that branch). I suppose since the unit could be different, it may be wrong to cast `start` first, so I guess casting later is correct?

Reply:
You are right, I had put in the line there because before we made `step` unit-aware, this was necessary. Now it's redundant, since the return statement has the cast anyway. I remove it now.

Reply:
Ah man, I messed up here I think :(. `delta / div` has to be a floating point number, but if `delta` is timedelta, it is not. That should break precision somewhere along the line. Since it happens anyway, I suppose we could cast `delta` to float directly, only that does currently not convert NaT to NaN, which is a bug...

Reply:
I understand what you mean, but is that really a problem? I have tried this test, that passes:
```python
    def test_datetime_precision(self):
        t1 = array('1969-06-06 00:00:00', dtype='M8[s]')
        dt = timedelta64(1, 's')
        N = 100
        t2 = t1 + dt*N
        assert_((linspace(t1, t2, N+1, dtype='M8[s]') == t1+dt*arange(N+1)).all())
```
Of course if `N` is so high that `N`  equally spaced samples cannot be drawn between `start` and `stop` with `dtype` precision, it won't work.

Reply:
Ohh no, I was wrong. There is a problem:
```python
In [20]: t1 = np.array('2019-11-12 08:00:00', 'M8[s]')                                     
In [21]: t2 = np.array('2019-11-12 08:00:01', 'M8[s]')                                     
In [23]: np.linspace(t1, t2, 5, dtype='M8[ms]')                                            
Out[23]: 
array(['2019-11-12T08:00:00.000', '2019-11-12T08:00:00.000',
       '2019-11-12T08:00:00.000', '2019-11-12T08:00:00.000',
       '2019-11-12T08:00:01.000'], dtype='datetime64[ms]')
```
But this can be taken care of by casting `start` and `stop` to `dtype` at the beginning. 
Is there anything else I am missing? I will in any case add these tests to the testsuite.

Reply:
Even then `t2 - t1` will be `numpy.timedelta64(1000,'ms')` if you have `num=33`: `(t2 - t1) / div` gives `numpy.timedelta64(29,'ms')`, but `(t2 - t1).astype(np.int64) / 34` (which gives a float value) is `29.41176470588235`, multiply that by 30 and you have a catastrophic failure?

Reply:
I see there's a problem, but I am sorry to say I cannot figure out what the expected behaviour of `linspace` in this case should be. Considering:
```ipython
In [15]: np.linspace(1, 10, 7, dtype=np.int64)                                             
Out[15]: array([ 1,  2,  4,  5,  7,  8, 10])
```
This datetime case yields, btw:
```ipython
In [8]: np.linspace(t1, t2, 33, dtype='M8[ms]')                                            
Out[8]: 
array(['2019-11-12T08:00:00.000', '2019-11-12T08:00:00.031',
       '2019-11-12T08:00:00.062', '2019-11-12T08:00:00.093',
       '2019-11-12T08:00:00.124', '2019-11-12T08:00:00.155',
       '2019-11-12T08:00:00.186', '2019-11-12T08:00:00.217',
       '2019-11-12T08:00:00.248', '2019-11-12T08:00:00.279',
       '2019-11-12T08:00:00.310', '2019-11-12T08:00:00.341',
       '2019-11-12T08:00:00.372', '2019-11-12T08:00:00.403',
       '2019-11-12T08:00:00.434', '2019-11-12T08:00:00.465',
       '2019-11-12T08:00:00.496', '2019-11-12T08:00:00.527',
       '2019-11-12T08:00:00.558', '2019-11-12T08:00:00.589',
       '2019-11-12T08:00:00.620', '2019-11-12T08:00:00.651',
       '2019-11-12T08:00:00.682', '2019-11-12T08:00:00.713',
       '2019-11-12T08:00:00.744', '2019-11-12T08:00:00.775',
       '2019-11-12T08:00:00.806', '2019-11-12T08:00:00.837',
       '2019-11-12T08:00:00.868', '2019-11-12T08:00:00.899',
       '2019-11-12T08:00:00.930', '2019-11-12T08:00:00.961',
       '2019-11-12T08:00:01.000'], dtype='datetime64[ms]')

In [12]: np.diff(np.linspace(t1, t2, 33, dtype='M8[ms]'))                                                                       
Out[12]: 
array([31, 31, 31, 31, 31, 31, 31, 31, 31, 31, 31, 31, 31, 31, 31, 31, 31,
       31, 31, 31, 31, 31, 31, 31, 31, 31, 31, 31, 31, 31, 31, 39],
      dtype='timedelta64[ms]')
```

I am sorry if I failed to get your point.

Reply:
No, the point is exactly that due to the step of 31 here, the intermediate values will be off by more than 1ULP for most values. This can always happen to some degree, but as is, for integers it behaves much better:
```
In [1]: np.diff(np.linspace(0, 1000, 33, dtype='int64'))                                                                
Out[1]: 
array([31, 31, 31, 32, 31, 31, 31, 32, 31, 31, 31, 32, 31, 31, 31, 32, 31,
       31, 31, 32, 31, 31, 31, 32, 31, 31, 31, 32, 31, 31, 31, 32])
```

Reply:
I think your previous solution behaved better/correctly here :/. We are bound to bake in some annoying cases (which is not nice, since once it is in, it is difficult to fix), but weird NaT handling seems better than this to me...

Reply:
Now I understand, thanks. So the options are:
* NaT handling consistent with NaN, but outputs may in many cases be more off than necessary. 
* We must explicitly disallow either start or stop being NaT. Output as precise as `dtype` allows. 

I find both these options annoying, but maybe not being able to call `linspace` with `NaT`'s are not that big a disadvantage?